### DISCUSSION TITLE
Blockly node

### DISCUSSION THREAD

**Discussion by phoddie** [phoddie/node-red-mcu#126] [Ideas]
@ralphwetzel raised the possibility of having the Node-RED Blockly node working on Node-RED MCU Edition. His original comment is [here](https://github.com/ralphwetzel/node-red-mcu-plugin/issues/50#issuecomment-2062170608) (ralphwetzel/node-red-mcu-plugin#50)

Having Blockly available could be interesting for Node-RED MCU in the classroom, something I know @mshioji has been actively exploring.

I guess this would be straightforward if the Blockly node includes the generated function source code in the JSON it exports.

Anyway.... I'm opening this issue to raise the profile of this idea and encourage others who may find this useful to share their thoughts.

**Comment by spillz-dxb** [comment]
Great , Looks like the start of making a stand alone visual programming tool for moddable . Looking forward to, however a migration from flow based would be hard , as we are used to node-red so much. What i found could be useful is you could call python packages , npm , java jars and c libraries . correct me if im wrong .If possible ,that would expand NR  access beyond npm packages.

**Comment by phoddie** [comment]
> What i found could be useful is you could call python packages , npm , java jars and c libraries . correct me if im wrong .If possible ,that would expand NR access beyond npm packages.

A microcontroller is not a computer. It cannot possibly provide Python, npm, Java, and C compatibility with a computer.

**Comment by ralphwetzel** [comment]
> I guess this would be straightforward if the Blockly node includes the generated function source code in the JSON it exports.

With reference to the [source of the node](https://github.com/bartbutenaers/node-red-contrib-blockly/blob/f017dbbc54708f2c454509a408bfb184150d35db/blockly.html#L1002-L1005), it should be available ... in `node.func`.

I haven't tested this, yet maybe it could be "enough" to tell `nodered2mcu.js` to process the `Blockly` node as a (standard) `function` node...

**Reply by phoddie** [reply]
That does look straightforward. Remapping in `nodered2mcu` is easy. I'll try to find some time to give that a try.

**Reply by phoddie** [reply]
@ralphwetzel – you are completely correct. It is just necessary to add a few lines to `nodered2mcu` to map the Blockly node to a Function node, and remove some of the bookkeeping data from the configuration.

Here are the lines to update in `prepareNode` if you want to try before this is pushed to the Moddable SDK repo.

```js
			case "Blockly":
				config.type = "function";
				delete config.workspaceXml;
				delete config.timeout;
				delete config.blocklyConfig;
				delete config.backpackContents;
				// fall through
			case "function": {
```

**Reply by ralphwetzel** [reply]
Working here! 👍
Didn't test any edge cases yet...

**Reply by phoddie** [reply]
@ralphwetzel – thanks for confirming that it works for you. I haven't explored the edges yet either. We will see.

**Reply by mshioji** [reply]
Hello @ralphwetzel and @phoddie,

 Wow, it's excellent! I also see it is working on MCU! (only a very simple blocks but...)
and I found it is already multi language supported including Japanese.^^
![image](https://github.com/phoddie/node-red-mcu/assets/59914628/86b87b6f-f554-4f72-a27b-188b75423db0)

Block Programming is a nice starting point for younger age, and Blockly node shows JS code as well, it is very helpful to learn native JavaScript code. it could be a very nice gateway. It is very impressive that this is also available in MCU programming!
